edgekv/src/datastore.rs

```
@@ -0,0 +1,835 @@
+        let index = DataIndex::InBuffer;
+        keys_dir_writer.insert(key, index);
+        Ok(())
+    }
+    pub fn remove(&self, key: &[u8]) -> Result<()> {
+        let mut keys_dir_writer = self
+            .keys
+            .write()
+            .map_err(|e| EdgeKVError::RWLockPoisonError(format!("{}", e)))?;
+        keys_dir_writer.remove(key);
+        Ok(())
+    }
+
+    pub fn clear(&self) -> Result<()> {
+        let mut keys_dir_writer = self
+            .keys
+            .write()
+            .map_err(|e| EdgeKVError::RWLockPoisonError(format!("{}", e)))?;
+        keys_dir_writer.clear();
+        Ok(())
+    }
+
+
+    //TODO change to use references
+    pub fn keys(&self) -> Vec<Vec<u8>> {
+        let keys_dir_reader = match self.keys.read() {
+            Ok(rdr) => rdr,
+            Err(_) => {
+                return vec![];
+            }
+        };
+
+        keys_dir_reader.iter().map(|(k, _)| k.clone()).collect()
+    }
+
+    //TODO change to use references
+    pub fn range<R>(&self, range : R) -> Vec<Vec<u8>> where R : RangeBounds<Vec<u8>> {
+        let keys_dir_reader = match self.keys.read() {
+            Ok(rdr) => rdr,
+            Err(_) => {
+                return vec![];
+            }
+        };
+        keys_dir_reader.range(range).map(|(k, _)| k.clone()).collect()
+    }
+
+    //TODO change to use references
+    pub fn prefix(&self, prefix: &Vec<u8>) -> Vec<Vec<u8>> {
+        let keys_dir_reader = match self.keys.read() {
+            Ok(rdr) => rdr,
+            Err(_) => {
+                return vec![];
+            }
+        };
+        keys_dir_reader
+            .range(prefix.clone()..)
+            .take_while(|(k, _)| k.starts_with(prefix))
+            .map(|(k, _)| k.clone())
+            .collect()
+    }
+
+
+    pub fn get(&self, key: &[u8]) -> Option<KeyDirEntry> {
+        let keys_dir_reader = match self.keys.read() {
+            Ok(rdr) => rdr,
+            Err(_) => {
```

What errors are possible here? should it be silently ignored?